src/test/script_tests.cpp

```diff
@@ -155,37 +159,59 @@ CMutableTransaction BuildSpendingTransaction(const CScript& scriptSig, const CSc
+            // Removing some random flags should not invalidate a valid script
+            flags = TrimFlags(~(scriptflags[0] | InsecureRandBits(FLAG_MAP_SIZE)));
+            BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey, &scriptWitness, flags, MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err), "Script unexpectedly failed with random flags " << std::to_string(flags) << ": " << message);
+        }
+        for (size_t i = 1; i < scriptflags.size(); i++)
+        {
+            // If the flags removed are minimal, adding a flag back should make the transaction invalid
+            BOOST_CHECK_MESSAGE(!VerifyScript(scriptSig, scriptPubKey, &scriptWitness, FillFlags(~scriptflags[i]), MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err), "Too many flags unset: " << message);
+        }
+    }
+    else {
+        for (size_t i = 0; i < FLAG_MAP_SIZE; i++)
+        {
+            // Adding a flag should not validate an invalid script
+            unsigned int flags = FillFlags(scriptflags[0] | (1U << i));
+            BOOST_CHECK_MESSAGE(!VerifyScript(scriptSig, scriptPubKey, &scriptWitness, flags, MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err), "Script unexpectedly passed with flag " << std::to_string(i) << " set: " << message);
+            // Adding some random flags should not validate an invalid script
+            flags = FillFlags(scriptflags[0] | InsecureRandBits(FLAG_MAP_SIZE));
+            BOOST_CHECK_MESSAGE(!VerifyScript(scriptSig, scriptPubKey, &scriptWitness, flags, MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err), "Script unexpectedly passed with random flags " << std::to_string(flags) << ": " << message);
+        }
+        for (size_t i = 1; i < scriptflags.size(); i++)
+        {
+            // If the flags used are minimal, removing a flag should make the transaction valid
+            // This test is skipped for certain CHECKSIG tests, so it is possible to demonstrate how a DERSIG or
+            // STRICTENC flag would terminate the script validation.
+            BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey, &scriptWitness, TrimFlags(scriptflags[i]), MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err) || (skip_flag_removal_test && err == SCRIPT_ERR_EVAL_FALSE), "Too many flags set: " << message);
+        }
     }
 
 #if defined(HAVE_CONSENSUS_LIB)
     CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
     stream << tx2;
-    int libconsensus_flags = flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
-    if (libconsensus_flags == flags) {
+    int libconsensus_flags = main_flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
+    if (expect || libconsensus_flags == (int)main_flags) {
```

Comment:
rebased and fixed a bug here. Without `expect ||`, all valid script tests were skipped due to the use of standardness flags. If a script is valid with standardness flags, removing these flags shouldn't make it invalid.

This also fixed an existing bug that many tests containing standardness flags (mostly STRICTENC) are unnecessarily skipped.